### DISCUSSION TITLE
Browser custom command not visible after update to 7.9.0 version

### DISCUSSION THREAD

**Discussion by NavyMint** [webdriverio/webdriverio#7196] [General]
Hey, dear wdio team :) One more question appeared after update to new 7.9.0 version

Previously we had custom command in wdio configuration and we used it everywhere within migrated page objects, but after update to 7.9.0 version custom commands are not visible in list of browser commands.

**Config file:**
```
 before: async () => {
        browser.addCommand('customParams', function () {
            return {
                downloadPath: `\\downloads`,
                relativeDownloadPath: 'downloads'
            };
        });

        browser.addCommand('qaId', async function(selector: string) {
            return $$(`[qaId=${selector}]`);
        }, true);
}
```

**wdio.d.ts**

```
declare namespace WebdriverIO {
    const typeOfCustomParams: {
        downloadPath: string;
        relativeDownloadPath: string;
    };

    interface Browser {
        customParams: () => typeof typeOfCustomParams;
        qaId: (selector: string) => ChainablePromiseElement<Promise<WebdriverIO.Element>>;
    }

    interface Element {
        qaId: (selector: string) => ChainablePromiseElement<Promise<WebdriverIO.Element>>;
    }
}
```

**spec**
```
browser.customCommand() or browser.qaId($'element') shows error 
Property 'customParams' does not exist on type 'Browser'
Property 'qaId' does not exist on type 'Browser'
```

Guys please suggest us how to use this custom command with Chainable return type, may be we need to add some additional changes in wdio.d.ts or config file?

Thank you in advance

**Comment by NavyMint** [comment]
@christian-bromann can you please take a look? may be its an issue or we just need to change something accordingly to new fixes with async

I also tried to change namespace to WebdriverIOAsync and 
```
    interface Browser {
        qaId: (selector: string) => Promise<ChainablePromiseElement<ElementArray>>;
    }

    interface Element {
        qaId: (selector: string) => Promise<ChainablePromiseElement<ElementArray>>;
    }
```

but also not working this way, custom command not visible in browser and element objects

**Comment by NavyMint** [comment]
we found a temporal workaround -> specifying any return type and then custing where its used but it looks ugly)) 
```
    interface Browser {
        qaId: (selector: string) => any;
    }

    interface Element {
        qaId: (selector: string) => any;
    }
```

**Comment by erwinheitzman** [comment]
```js
        browser.addCommand('qaId', async function(selector: string) {
            return $$(`[qaId=${selector}]`);
        }, true);
```
By passing true, you assign it to the Element scope instead of the Browser scope.
So assuming based on the info you have shared, the error would in fact be correct.

Have you perhaps not shared everything?

**Reply by NavyMint** [reply]
@erwinheitzman 
From Element scope its also not visible, I guess the reason in using ChainablePromise return type in file **wdio.d.ts**
When Im adding it - custom command becomes not recognizable for browser/element objects
When Im using in wdio.d.ts. return type as any - I can see my custom commands

**Reply by erwinheitzman** [reply]
It becomes visible because you have defined both in your `wdio.d.ts` yes, but in reality the method is only available on the Element scope.

So in short:
- It shows for both because you have defined these types
- It is only accessible on the Element scope in reality
- There seems to be an issue with your typings

Could you figure out what exactly is triggering the issue and let us know?

**Reply by erwinheitzman** [reply]
Another thing is that when using `$$` the I believe that the return type should be `ElementArray` (maybe `Promise<ElementArray>`) which is exposed by the async webdriverio types. I don't know what the `ChainablePromiseElement ` is doing here because this is used by the commands themselves and not the return type.

**Reply by NavyMint** [reply]
I've tried to simplify code and check once again
**wdio.conf.ts**
![image](https://user-images.githubusercontent.com/26191692/127477729-ae55340d-b83a-4d92-ba70-d394e6e2fccb.png)


**wdio.d.ts**
![image](https://user-images.githubusercontent.com/26191692/127477643-d0c96d5e-01c8-411e-bcff-61bf339e8bf1.png)


**test case**
![image](https://user-images.githubusercontent.com/26191692/127477898-6cbc3f75-8a48-45a4-bf3c-e3d6f9a8456f.png)

**Reply by NavyMint** [reply]
I've tried to use this ChainablePromise as a return type coz all elements automatically showing return type Chainable after upgrade to 7.9 version

**Reply by erwinheitzman** [reply]
Maybe you need to overwrite the ChainablePromiseElement instead of Element.
However you did say that changing the return type of the dataId method, fixes the issue right?

It might be that we broke something because overwriting the Element interface should work

**Reply by erwinheitzman** [reply]
I'll have a check myself later

**Reply by NavyMint** [reply]
yep, as a temp workaround we can specify in wdio.d.ts return type any and then in the test cases or page objects cast to chainable type - in this way it works. But overall its not too convinient in practice coz we need to force this casting in all places where custom commands used ((

Thank you for looking into it with me, please let me know in case you find out something, and in case its an issue we will be very thankful for any row estimations when we can expect a fix :)

**Reply by erwinheitzman** [reply]
<img width="633" alt="Screenshot 2021-07-29 at 15 10 28" src="https://user-images.githubusercontent.com/15839059/127498230-8b7556a5-f04b-4957-8060-d7f040346efb.png">
<img width="931" alt="Screenshot 2021-07-29 at 15 10 12" src="https://user-images.githubusercontent.com/15839059/127498235-d9be3c33-cba8-485a-8c60-c981e7bac7d8.png">

The following works just fine, I think the issue is the import in your `wdio.d.ts`.
Hope this helps 😉

**Reply by NavyMint** [reply]
yeeeeeeeeeeaaah, thanks a mile! Its working this way!! :))
